### PR TITLE
feature/controlling the proportion of the words in the sentence that get transformed

### PR DESCRIPTION
**Pull request by RakshitKhajuria**
## Checklist:

- [x] I've added Google style docstrings to my code.
- [x] I've used `pydantic` for typing when/where necessary.
- [x] I have linted my code
- [x] I have added tests to cover my changes.

## Structure of the config for controlling the proportion of the test sentences
```
harness.configure(
{
 'tests': {'defaults': {'min_pass_rate': 0.65},
           'robustness': {'uppercase': {'min_pass_rate': 0.66, 'prob': 0.50},
                          'lowercase':{'min_pass_rate': 0.60,'prob': 0.70},
                          'titlecase':{'min_pass_rate': 0.60,'prob': 0.80},
                          'add_punctuation':{'min_pass_rate': 0.60,'prob': 0.30},
                        }
          }
 }
 )
```

### CODE REVIEW

File: nlptest/transform/robustness.py [hunk 1]
```diff
@@ -88,79 +88,103 @@ async def async_run(cls, sample_list: List[Sample], model: ModelFactory, **kwarg
...
-
-
 class UpperCase(BaseRobustness):
     alias_name = "uppercase"
 
     @staticmethod
-    def transform(sample_list: List[Sample]) -> List[Sample]:
-        """Transform a list of strings with uppercase robustness
+    def transform(sample_list: List[Sample], prob: float = None) -> List[Sample]:
+        """Transform a list of strings with uppercase robustness.
```
Comment (JulesBelveze) [inline nlptest/transform/robustness.py:96]: By setting the default value to `1.0` you can directly have `num_transform_words = int(prob * len(words))` and don't need to check if `prob is None` 😃

File: nlptest/transform/robustness.py [hunk 2]
```diff
@@ -88,79 +88,103 @@ async def async_run(cls, sample_list: List[Sample], model: ModelFactory, **kwarg
...
         Args:
             sample_list: List of sentences to apply robustness.
+            prob: Probability of perturbing each word. If None, all words are perturbed.
         Returns:
-            List of sentences that uppercase robustness is applied.
+            List of sentences with uppercase robustness applied.
```
Comment (JulesBelveze) [inline nlptest/transform/robustness.py:101]: Please add the parameters and return values types in the docstring

**Review by JulesBelveze — CHANGES_REQUESTED** [review]
Cool feature! 🤓 
The two comments I left apply to all transformation can you please change the `prob` default value everywhere? 
Also, would be great to add some tests in which we pass the parameter `prob` to the `transform` method

File: nlptest/transform/robustness.py (hunk 1 above)
Comment (Prikshit7766) [reply nlptest/transform/robustness.py:96]: ya sure

Comment (RakshitKhajuria) [reply nlptest/transform/robustness.py:96]: Alright !

File: nlptest/utils/custom_types/sample.py [hunk 3]
```diff
@@ -345,7 +345,7 @@ class BaseQASample(BaseModel):
     def __init__(self, **data):
         super().__init__(**data)
 
-    def transform(self, func, params, prob=None, **kwargs):
+    def transform(self, func, params, prob, **kwargs):
         sens = [self.original_question, self.original_context]
         self.perturbed_question, self.perturbed_context = func(sens, prob, **params, **kwargs)
         self.category = func.__module__.split('.')[-1]
```
Comment (JulesBelveze) [inline nlptest/utils/custom_types/sample.py]: Can you add typing and docstring here please?

File: nlptest/utils/custom_types/sample.py [hunk 4]
```diff
@@ -520,7 +520,7 @@ def _is_eval(self) :
             results = metric.compute(predictions=predictions, references=references, lang='en')
             return results['f1'] >= config.get('threshold', 0.50), results['f1']
     
-    def transform(self, func, params, prob=None, **kwargs):
+    def transform(self, func, params, prob, **kwargs):
         """"""
         sens = [self.original]
         self.test_case= func(sens, prob, **params, **kwargs)[0]
```
Comment (JulesBelveze) [inline nlptest/utils/custom_types/sample.py]: Same here

File: nlptest/transform/robustness.py [hunk 5]
```diff
@@ -103,13 +103,13 @@ def transform(sample_list: List[Sample], prob: float = None) -> List[Sample]:
         for idx, sample in enumerate(sample_list):
             if isinstance(sample, str):
                 words = sample.split()
-                num_transform_words = int(prob * len(words)) if prob is not None else len(words)
+                num_transform_words = num_transform_words = int(prob * len(words))
                 transformed_indices = random.sample(range(len(words)), num_transform_words)
```
Comment (JulesBelveze) [inline nlptest/transform/robustness.py]: Hmmm why do we have `num_transform_words = num_transform_words = ...`?

**Review by JulesBelveze — CHANGES_REQUESTED** [review]
Minor changes needed, please add some simple tests and then we good to go 🚀 
Thanks!

File: nlptest/transform/robustness.py (hunk 5 above)
Comment (JulesBelveze) [reply nlptest/transform/robustness.py]: This is also in a bunch of other places

**Review by JulesBelveze — APPROVED** [review]
LGTM 🤓

**Pull request merged**